lightgbm/src/main/scala/com/microsoft/azure/synapse/ml/lightgbm/LightGBMBase.scala

```diff
@@ -374,251 +376,288 @@ trait LightGBMBase[TrainedModel <: Model[TrainedModel]] extends Estimator[Traine
+      case _ => throw new IllegalArgumentException("Unknown row data type to push")
     }
-    val f = Future {
-      var emptyTaskCounter = 0
-      val hostAndPorts = ListBuffer[(Socket, String)]()
-      val hostToMinPartition = mutable.Map[String, String]()
-      if (getUseBarrierExecutionMode) {
-        log.info(s"driver using barrier execution mode")
 
-        def connectToWorkers: Boolean = handleConnection(driverServerSocket, log,
-          hostAndPorts, hostToMinPartition) == Finished || connectToWorkers
+    val numInitScoreClasses =
+      if (get(initScoreCol).isEmpty) 0
+      else if (dataframe.schema(getInitScoreCol).dataType == VectorType)
+        firstRow.getAs[DenseVector](getInitScoreCol).size
+      else 1
 
-        connectToWorkers
-      } else {
-        log.info(s"driver expecting $numTasks connections...")
-        while (hostAndPorts.size + emptyTaskCounter < numTasks) {
-          val connectionResult = handleConnection(driverServerSocket, log, hostAndPorts, hostToMinPartition)
-          if (connectionResult == ConnectionState.EmptyTask) emptyTaskCounter += 1
-        }
-      }
-      // Concatenate with commas, eg: host1:port1,host2:port2, ... etc
-      val hostPortsList = hostAndPorts.map(_._2).sortBy(hostPort => {
-        val host = hostPort.split(":")(0)
-        Integer.parseInt(hostToMinPartition(host))
-      })
-      val allConnections = hostPortsList.mkString(",")
-      log.info(s"driver writing back to all connections: $allConnections")
-      // Send data back to all tasks and helper tasks on executors
-      sendDataToExecutors(hostAndPorts, allConnections)
-      closeConnections(log, hostAndPorts, driverServerSocket)
-    }
-    val host = ClusterUtil.getDriverHost(spark)
-    val port = driverServerSocket.getLocalPort
-    log.info(s"driver waiting for connections on host: $host and port: $port")
-    (host, port, f)
+    measures.markColumnStatisticsStop()
+    (numCols, numInitScoreClasses)
   }
 
   /**
     * Inner train method for LightGBM learners.  Calculates the number of workers,
     * creates a driver thread, and runs mapPartitions on the dataset.
     *
-    * @param dataset    The dataset to train on.
+    * @param dataframe The dataset to train on.
+    * @param trainingParams The training parameters.
+    * @param numCols The number of feature columns.
+    * @return The serialized Dataset reference and an array of partition counts.
+    */
+  protected def calculateRowStatistics(dataframe: DataFrame,
+                                       trainingParams: BaseTrainParams,
+                                       numCols: Int,
+                                       measures: InstrumentationMeasures): (Array[Row], Array[Long]) = {
+    measures.markRowStatisticsStart()
+
+    // Get the row counts per partition
+    measures.markRowCountsStart()
+    val indexedRowCounts: Array[(Int, Long)] = dataframe
+      .select(getLabelCol)
+      .rdd
```

Comment:
nit: you might be able to avoid rdd api by using some api on the task context or something like this to get the partition number

Reply:
not sure what you mean. this is getting a count of the number of rows in each partition by partition id.  there is no information about partition row count in task context.  And if there is, PLEASE show me. :)

Reply:
I'll also need that information for VW bindings. can we move it into a util... and given anybody finds a better way of doing it, update it there?

Reply:
> I'll also need that information for VW bindings. can we move it into a util... and given anybody finds a better way of doing it, update it there?

You need what information exactly?  Just so I'm clear on what I'm turning into a util. :)

Reply:
I need to know the number of rows per partition to trigger all reduce.

Reply:
done, added a util in core ClusterUtil